### PR TITLE
Refactor Ground Control and default Source sublanguage feature

### PR DESCRIPTION
**Pull request by Aulud**
## Refactor Ground Control and default Source sublanguage feature
Summary: Major refactor including changes to address #1212, #1286 and #1320.

### Related PRs
Please review and approve this PR in tandem with the PR on cadet: https://github.com/source-academy/cadet/pull/627. **This PR is complete**.

### Description
This PR refactors both the Ground Control and the default Source sublanguage feature, with small class components replaced with their equivalent function components using React hooks and other general code quality improvements and bug fixes.

The main UI/UX change involves the creation of a control panel in the Ground Control, above the datagrid containing the Academy's assessments. The control panel now contains the default Source sublanguage selection menu, which was previously (unsuitably) located in the navigation bar. Additionally, the large central dropzone for uploading assessment files is now hidden by default to de-clutter the page, in favour of a button in the control panel that toggles whether the dropzone is displayed.

### Changelog
- Rename where appropriate to distinguish a Source sublanguage from its components (chapter and variant)
  - Remove duplicate sublanguage constants and types from application types and across related components
- Remove unnecessary route props and `withRouter` call in default sublanguage selection container component
- Add missing backend saga for `FETCH_SUBLANGUAGE` action to fetch default Source sublanguage
- Update default Source sublanguage endpoints (see https://github.com/source-academy/cadet/pull/627)
  - Route to retrieve sublanguage: `GET /settings/sublanguage`
  - Route to update sublanguage: `PUT /settings/sublanguage`
- Add warning notifications for failed requests to default Source sublanguage endpoints
- Refactor default sublanguage selection component as a function component using React hooks
- Add confirmation prompt for updating the default Source sublanguage (addresses #1212)
- Fetch default Source sublanguage when accessing the Playground without a share link (addresses #1320)
  - This applies to all users, including those that are not logged into the Academy
- Group variant Source sublanguages in default sublanguage selection menu into a separate sub-menu (see #1354)
- Relocate default sublanguage selection component to control panel in Ground Control (addresses #1286)
- Fix repetitive flickering of Ground Control datagrid on every render
  - Ground Control will now only fetch the list of assessments from the backend if not previously fetched
  - A button has been added to the control panel to reload the list of assessments from the backend
- Fix Ground Control unnecessarily recreating datagrid component on re-renders
- Fix spurious horizontal scrollbar in datagrid body at certain widths
- Fix Ground Control dropzone accepting multiple files at once despite only being capable of uploading a single assessment
  - The dropzone will now reject attempts to upload multiple files with a warning notification
- Fix Ground Control crashing from edit cells accepting empty or invalid dates for assessments
  - Entering an empty or invalid date in the edit cell will reset its value to the current opening/closing date for that assessment with a warning notification
- Remove redundant datagrid props and column props
- Allow native datagrid filtering and sorting of assessments by opening and closing dates
- Refactor all Ground Control datagrid cell renderers as function components using React hooks
  - Redundant interfaces were removed and remaining interfaces were converted to their equivalent types
- Extract out inline CSS of Ground Control dropzone into `_groundcontrol.scss`
  - The list of classes to apply for the dropzone is now dynamically computed off its state
- Refactor Ground Control dropzone as a function component using React hooks
  - States required for the dropzone has been extracted out of Ground Control to avoid unnecessary prop chaining
- Hide Ground Control dropzone by default in a collapsed region with a button to toggle display of the full region
- Improve UI of rendered cells in the datagrid and assessment upload confirmation prompts in the dropzone
- Update snapshots for navigation bar
- Add tests for `FETCH_SUBLANGUAGE` backend saga

### Verification instructions
WIP

Last updated 9 Aug 2020, 10:45PM

### CODE REVIEW

File: src/pages/academy/groundControl/subcomponents/DefaultChapterSelect.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        <div className={Classes.DIALOG_FOOTER_ACTIONS}>
+          {controlButton('Cancel', null, handleCloseDialog, { minimal: false })}
+          {controlButton('Confirm', null, handleConfirmDialog, {
+            minimal: false,
+            intent: Intent.DANGER
+          })}
+        </div>
+      </div>
+    </Dialog>
+  );
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/DefaultChapterSelect.tsx:106]: Could use DialogHelper here

File: src/pages/academy/groundControl/subcomponents/DefaultChapterSelectContainer.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    },
+    dispatch
+  );
+
+const DefaultChapterSelectContainer = connect(
+  mapStateToProps,
+  mapDispatchToProps
+)(DefaultChapterSelect);
+
+export default DefaultChapterSelectContainer;
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/DefaultChapterSelectContainer.ts:30]: Boilerplate file :(

File: src/pages/academy/groundControl/subcomponents/GroundControlDeleteCell.tsx [hunk 3]
```diff
@@ -1,60 +1,64 @@
...
-  private handleOpenDialog = () => this.setState({ dialogOpen: true });
-  private handleDelete = () => {
-    const { data } = this.props;
-    this.props.handleDeleteAssessment(data.id);
-    this.handleCloseDialog();
-  };
-}
+        </div>
+      </Dialog>
+    </>
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/GroundControlDeleteCell.tsx:60]: Another place to use the DialogHelper

File: src/pages/academy/groundControl/subcomponents/GroundControlDropzone.tsx [hunk 4]
```diff
@@ -1,146 +1,139 @@
...
+          intent: Intent.DANGER
+        })}
+        {controlButton('No', IconNames.CROSS, handleCancelForceUpdate, {
+          minimal: false
+        })}
       </div>
-    );
-  };
+    ),
+    [handleCancelForceUpdate, handleConfirmForceUpdate]
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/GroundControlDropzone.tsx:104]: This weird prompt thing could also be done with DialogHelper

File: src/pages/academy/groundControl/subcomponents/GroundControlEditCell.tsx [hunk 5]
```diff
@@ -1,103 +1,119 @@
...
-      this.setState({ closeAt: selectedDate });
-    }
-  };
+  const dateInput = (
+    <DateInput
+      formatDate={handleFormatDate}
+      onChange={handleDateChange}
+      onError={handleDateError}
+      parseDate={handleParseDate}
+      placeholder={'MM/DD/YYYY, HH:mm:ss (AM/PM) or select a date'}
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/GroundControlEditCell.tsx:80]: American date format :confused:

File: src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx [hunk 6]
```diff
@@ -1,70 +1,73 @@
...
-  private handleCloseDialog = () => this.setState({ dialogOpen: false });
-  private handleOpenDialog = () => this.setState({ dialogOpen: true });
-  private handleDelete = () => {
-    const { data } = this.props;
-    this.props.handlePublishAssessment(!data.isPublished, data.id);
-    this.handleCloseDialog();
-  };
-}
+        </div>
+      </Dialog>
```
Comment (angelsl) [inline src/pages/academy/groundControl/subcomponents/GroundControlPublishCell.tsx:68]: Another place to use DialogHelper

File: src/pages/academy/groundControl/subcomponents/GroundControlEditCell.tsx (hunk 5 above)
Comment (Aulud) [reply src/pages/academy/groundControl/subcomponents/GroundControlEditCell.tsx:80]: Unfortunately this is the default JS date format - I'll look for a way to configure this.

**Pull request merged**